src/transformers/trainer_callback.py

```diff
@@ -563,6 +564,20 @@ def on_train_begin(self, args, state, control, **kwargs):
             args.evaluation_strategy != IntervalStrategy.NO
         ), "EarlyStoppingCallback requires IntervalStrategy of steps or epoch"
 
+    def on_train_end(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
+
+        checkpoints_sorted = state._sorted_checkpoints(use_mtime=True, output_dir=state.get_run_dir())
+
+        if (
+            state.best_model_checkpoint is not None
+            and state.args.save_total_limit == 1
+            and checkpoints_sorted[-1] != self.state.best_model_checkpoint
+        ):
+            checkpoints_to_be_deleted = checkpoints_sorted[-1]
+            for checkpoint in checkpoints_to_be_deleted:
+                logger.info(f"Deleting older checkpoint [{checkpoint}] due to args.save_total_limit")
+                shutil.rmtree(checkpoint)
+
```

Comment:
The fix should be more general I think, and at the end of the `train` method itself.